src/foam/nanos/auth/InvoiceHistoryAuthorizer.js

```diff
@@ -0,0 +1,91 @@
+foam.CLASS({
+  package: 'foam.nanos.auth',
+  name: 'InvoiceHistoryAuthorizer',
+  implements: ['foam.nanos.auth.Authorizer'],
+
+  documentation: `AuthorizableAuthorizer which checks if user has access to the invoice history record`,
+
+  javaImports: [
+    'foam.dao.DAO',
+    'foam.dao.ArraySink',
+    'net.nanopay.invoice.model.Invoice',
+    'foam.dao.history.HistoryRecord',
+    'static foam.mlang.MLang.*',
+    'foam.nanos.auth.AuthorizationException',
+  ],
+
+  methods: [
+    {
+      name: 'authorizeOnCreate',
+      javaCode: `
+        User user = (User) x.get("user");
+        if ( user == null ) {
+          throw new foam.nanos.auth.AuthenticationException("User not found.");
+        } else if ( ! "admin".equalsIgnoreCase(user.getGroup()) && ! "system".equalsIgnoreCase(user.getGroup())) {
+          throw new AuthorizationException();
+        }
+      `
+    },
+    {
+      name: 'authorizeOnRead',
+      javaCode: `
+        DAO invoiceDAO = (DAO) x.get("invoiceDAO");
+        User user = (User) x.get("user");
+
+        ArraySink invoiceArraySink = (ArraySink) invoiceDAO.where(
+          EQ(Invoice.ID, ((HistoryRecord) obj).getObjectId())
+        ).select(new ArraySink());
+        Invoice invoice = (Invoice) invoiceArraySink.getArray().get(0);
+        if ( (invoice.getPayeeId() != user.getId()) && (invoice.getPayerId() != user.getId()) ){
+          throw new AuthorizationException();
+        }
+      `
+    },
+    {
+      name: 'authorizeOnUpdate',
+      javaCode: `
+        User user = (User) x.get("user");
+        if ( user == null ) {
+          throw new foam.nanos.auth.AuthenticationException("User not found.");
+        } else if ( ! "admin".equalsIgnoreCase(user.getGroup()) && ! "system".equalsIgnoreCase(user.getGroup())) {
+          throw new AuthorizationException();
+        }
+      `
+    },
+    {
+      name: 'authorizeOnDelete',
+      javaCode: `
+        User user = (User) x.get("user");
+        if ( user == null ) {
+          throw new foam.nanos.auth.AuthenticationException("User not found.");
+        } else if ( ! "admin".equalsIgnoreCase(user.getGroup()) && ! "system".equalsIgnoreCase(user.getGroup())) {
+          throw new AuthorizationException();
+        }
+      `
+    },
+    {
+      name: 'checkGlobalRead',
+      javaCode: `
+        String permission = "invoiceHistoryDAO.read.*";
+        AuthService authService = (AuthService) x.get("auth");
+        try {
```

Comment:
I don't think using a try/catch here makes sense. We don't expect the `auth` service to throw an exception, we expect it to return true or false, depending on whether the user has the permission or not. If an exception is thrown in it then we should let it bubble up since we're not explicitly trying to do error handling here.

The code for `checkGlobalRead` and remove should be something like:

```Java
AuthService authService = (AuthService) x.get("auth");
return authService.check(x, "invoicehistory.read.*");
```